poc/prg.sage

```diff
@@ -99,6 +99,45 @@ class PrgSha3(Prg):
     def next(self, length: Unsigned) -> Bytes:
         return self.shake.read(length)
 
+# PRG based on a circular collision-resistant hash function from fixed-key AES.
+class PrgFixedKeyAes128(Prg):
+    # Associated parameters
+    SEED_SIZE = 16
+
+    def __init__(self, seed, custom, binder):
+        self.length_consumed = 0
+
+        # Use SHAKE128 to derive a key from the binder and customization strings.
+        # Note that the AES key does *not* need to be secret from any party, it just
+        # has to be random for each client, which is ensured by clients choosing
+        # nonces at random.
+        # In practice, this step can be cached across PRG evaluations with many
+        # different seeds.
+        shake = cSHAKE128.new(custom=custom)
+        shake.update(binder)
+        key = shake.read(16)
+        self.cipher = AES.new(key, AES.MODE_ECB)
+        # Save seed to be used in `next`.
+        self.seed = seed
+
+    def next(self, length: Unsigned) -> Bytes:
+        offset = self.length_consumed % 16
+        new_length = self.length_consumed + length
+        blocks = range(int(self.length_consumed / 16), int(new_length / 16) + 1)
+        self.length_consumed = new_length
+
+        # hash_block is a circular correlation-robust hash function.
+        # See https://eprint.iacr.org/2019/074.pdf pp. 18-19.
+        def hash_block(input):
```

Comment:
nit: Rather than define this inline, I think it would make more sense as a helper function defined in the main scope.